code/modules/reagents/machinery/chem_master.dm

```diff
@@ -136,418 +131,418 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
-				dat += "No info. Please contact technical support if you believe this is an error."
-			else
-				dat += "[condi ? "Condiment" : "Chemical"] information:<BR><BR>Name:<BR>[reagent.name]<BR><BR>Description:<BR>[reagent.description]<BR><BR>"
-				if(!condi)
-					if(istype(reagent, /datum/reagent/blood))
-						dat += "Blood type: [reagent.data["blood_type"] || "Unknown"]<BR>Blood DNA: [reagent.data["blood_DNA"] || "Unable to determine"]<BR><BR>"
-					dat += "Density:<BR>[reagent.density]<BR><BR>Specific heat capacity:<BR>[reagent.specheatcap]<BR><BR><BR>"
-				dat += "<A href='?src=\ref[src];main=1'>(Back)</A>"
-
-			dat = jointext(dat,"")
-			var/datum/browser/popup = new(usr, "[windowtype]", "[name]", 585, 400, src)
-			popup.set_content(dat)
-			popup.open()
-			return 1
-
-		else if(href_list["add"])
-			var/id = href_list["add"]
+		if(href_list["beaker_addall"])
 			var/amount
+
 			if(href_list["amount"])
 				amount = text2num(href_list["amount"])
-			else if(href_list["percent"])
-				amount = R.get_reagent_amount(id) * text2num(href_list["percent"]) / 100
 			if(isnull(amount) || amount < 0)
 				return
-			R.trans_id_to(src, id, amount)
+
+			if(clear_reagents)
+				src.reagents = buffer
+			else
+				src.reagents = storage
+
+			R.trans_to(src, amount)
 			src.updateUsrDialog()
 			return 1
 
-		else if(href_list["addall"])
+		if(href_list["beaker_addcustom"])
+			var/id = href_list["beaker_addcustom"]
+
+			useramount = input("Select the amount to transfer.", 30, useramount) as num
+			useramount = isgoodnumber(useramount)
+			src.Topic(null, list("amount" = "[useramount]", "beaker_add" = "[id]"))
+
+			src.updateUsrDialog()
+			return 1
+
+		if(href_list["beaker_add"])
+			var/id = href_list["beaker_add"]
 			var/amount
+
 			if(href_list["amount"])
 				amount = text2num(href_list["amount"])
-			else if(href_list["percent"])
-				amount = R.total_volume * text2num(href_list["percent"]) / 100
 			if(isnull(amount) || amount < 0)
 				return
-			R.trans_to(src, amount)
+
+			if(clear_reagents)
+				src.reagents = buffer
+			else
+				src.reagents = storage
+
+			R.trans_id_to(src, id, amount)
 			src.updateUsrDialog()
 			return 1
 
-		else if(href_list["remove"])
-			var/id = href_list["remove"]
+		if(href_list["eject"])
+			detach()
+			if(clear_reagents)
+				if(buffer && buffer.total_volume)
+					buffer.remove_all(buffer.total_volume)
+				if(storage && storage.total_volume)
+					storage.remove_all(storage.total_volume)
+			src.updateUsrDialog()
+			return 1
+
+	// Storage
+	if(!clear_reagents)
+		if(href_list["storage_addall"])
 			var/amount
+
 			if(href_list["amount"])
 				amount = text2num(href_list["amount"])
-			else if(href_list["percent"])
-				amount = reagents.get_reagent_amount(id) * text2num(href_list["percent"]) / 100
 			if(isnull(amount) || amount < 0)
 				return
-			if(mode)
-				reagents.trans_id_to(beaker, id, amount)
-			else
-				reagents.remove_reagent(id, amount)
+
+			if(storage_mode == 1)
```

Comment:
You could use defines as well as a switch statement here and in the following lines.